nonmember_subscript_operator/main.tex

```diff
@@ -125,6 +126,68 @@ \subsection{Compatiblity layer}
 
 \end{lstlisting}
 
+\subsection{Interpolated access}
+Subscripting a random-access range with non-integral types can have interesting use cases,
+like interpolating between the elements at the nearest smaller and larger integral index of the subscript:
+\smallskip\begin{lstlisting}
+auto operator[](std::ranges::random_access_range auto r, double index) {
+    double integ;
+    const double frac = std::modf(index, &integ);
+    return std::lerp(r[integ], r[integ + 1], frac);
+}
+std::vector v{0.0, 1.0, 4.5, 7.8};
+double value = v[1.6]; // value == 3.1
+\end{lstlisting}
+However, adding such an overload in practice might be dangerous because the floating-point index might truncate to a range's integral subscript operator in case the floating point overload is not visible, or change the behavior of an existing floating point subscription relying on the truncation.
+This issue could be worked around using a custom type which cannot truncate into an integral, allowing to e.g. also specify the used interpolation:
+\smallskip\begin{lstlisting}
+std::vector v{0.0, 1.0, 4.5, 7.8};
+double value = v[1.6_linear]; // value == 3.1
+\end{lstlisting}
+
+With the adoption of P2128, we could even allow interpolation on higher dimensional objects, like the proposed `std::mdspan` or `std::mdarray`.
+\smallskip\begin{lstlisting}
+std::mdarray<double, N, M, K> a = ...;
+double value = a[7.5, 8.9, 76.4]; // trilinear interpolation
+\end{lstlisting}
+
+While we are aware that such a usage might be niche or a separate function for interpolated access could be arguably better, we would like to give library writes the necessary machinery to implement such a feature within their domains if they desire so.
+
```

Comment:
It will be really interesting what EWG thinks about the interpolated access example (and the simd-example for that matter), as people stated that subscripting should always offer reference semantics...

Reply:
I love this feature so much, that I am probably blind to the critique. @mattkretz also did not like it a lot, especially the `float` subscript. Being a bit of a 3D graphics fan, I am used to stuff being interpolated all the time, so this feels natural to me. But sure, if the use case is to controversal, we can also drop it.

Reply:
Is there any precedent for a `float`-subscript in 3D graphics? Based on my albeit limited GLSL experience, I'm inclined to believe that whilst data is often interpolated, such an access is never expressed via subscripting...

Reply:
Hmm, looking that up now you are probably right:
* In OpenGL/GLSL there is the `texture(texObj, coords...)` function.
* In D3D/HLSL there is the `texObj.Sample(sampler, coords...)` member function.
* In Cg, there is the `tex2D(sampler, coords)` function.

So it seems there is no direct precedence in shading languages for interpolation using subscripting. Such functionality is indeed provided by functions.

Reply:
The question for me now becomes whether interpolation on subscription is even a bad idea in niche domains, domain specific libraries or closed internal code bases. I agree that such a feature does not belong into the standard library. But could there still be interesting applications in such limited scenarios?